### PR TITLE
Move console foreign handler to its own plugin

### PR DESCRIPTION
**Pull request by jasongrout**
This makes the foreign cell handler less coupled to the console, and shows a pattern for doing something like #5699 from a separate extension.

From a user's perspective, this should not change anything.

### CODE REVIEW

File: packages/console/src/foreign.ts [hunk 1]
```diff
@@ -145,7 +146,8 @@ export class ForeignHandler implements IDisposable {
    * Create a new code cell for an input originated from a foreign session.
    */
   private _newCell(parentMsgId: string): CodeCell {
-    let cell = this._factory();
+    let cell = this.parent.createCodeCell();
+    cell.addClass(FOREIGN_CELL_CLASS);
     this._cells.set(parentMsgId, cell);
```
Comment (BoPeng) [inline packages/console/src/foreign.ts]: This is the problem that I talked about. `_cells` is private to `ForeignHandler`, which is a map of `parentMsgId` and (reference to) `CodeCell` so the `parentMsgId` is only seen by the `ForeignHandler`, not by `console/widget` (whose `_cells` is a list of cells).

When both `ForeignHandler` and `TransientHandler` is enabled, the `ForeignHandler` will first get `execute_input` and create a cell, but `TransientHandler` does not know this cell has the same `parentMsgId` as the `transient_display_data` and will create a new cell.

So perhaps the mechanism of storing and querying `parentMsgId` need to be moved to console/widget.

Comment (BoPeng) [reply packages/console/src/foreign.ts]: Another problem with keeping `_cells` in `ForeignHandler` is that if the core or another extension decides to remove some cells (e.g. to keep the use of memory in check if the console gets really long), the `ForeignHandler` will have invalid references in its `_cells`.

Comment (BoPeng) [reply packages/console/src/foreign.ts]: @jasongrout Do you have an easy fix to the problem I mentioned or do you want me to work on a fix?

Comment (jasongrout) [reply packages/console/src/foreign.ts]: Do you want to try out what I just pushed? Does that look better to you?

Comment (BoPeng) [reply packages/console/src/foreign.ts]: Yes, it does look better. I will try it tonight by implementing my `TransientHandler` against it.

Comment (BoPeng) [reply packages/console/src/foreign.ts]: I can confirm that the PR works perfectly for my purpose and can be merged as soon as the tests are fixed.

Comment (jasongrout) [reply packages/console/src/foreign.ts]: Thanks again for patiently explaining the issues you saw. Very good points!

**Comment by jasongrout** [comment]
Thanks for the review and suggestions, @BoPeng

**Pull request merged**